File: packing_assistant/runtime/threads.py
"""Codex-app threads: one session per thread, parallel across threads.

Same session stays serial (Scheduler lock). /new and /bg get a new session_id.
"""

from __future__ import annotations

import json
import os
import time
from concurrent.futures import Future, ThreadPoolExecutor
from dataclasses import asdict, dataclass, field
from pathlib import Path
from threading import Lock
from typing import Any, Dict, List, Optional
from uuid import uuid4

_ROOT = Path(__file__).resolve().parents[2]
_DIR = _ROOT / "demo" / "out" / "_threads"
_LOCK = Lock()
_POOL: Optional[ThreadPoolExecutor] = None
_FUTS: Dict[str, Future] = {}
_CANCEL: set = set()
_BOOT = time.time()
MAX_MESSAGES = 400


@dataclass
class CivilThread:
    thread_id: str
    session_id: str
    title: str = ""
    skill: str = ""
    state: str = "idle"
    confirm: bool = False
    last_text: str = ""
    last_reply: str = ""
    hitl_pending: bool = False
    wrote: bool = False
    artifacts: List[str] = field(default_factory=list)
    error: str = ""
    created_at: float = field(default_factory=time.time)
    updated_at: float = field(default_factory=time.time)
    cancel_requested: bool = False
    n_messages: int = 0

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


def _pool() -> ThreadPoolExecutor:
    global _POOL
    if _POOL is None:
        from packing_assistant.runtime.civil_config import load_config

        n = load_config().max_parallel
        _POOL = ThreadPoolExecutor(max_workers=n, thread_name_prefix="civil-th")
    return _POOL


def _path(thread_id: str) -> Path:
    safe = "".join(ch if ch.isalnum() or ch in "-_" else "_" for ch in (thread_id or ""))[:40]
    return _DIR / f"{safe or 't'}.json"


def _atomic_write(path: Path, text: str) -> None:
    tmp = path.with_suffix(path.suffix + f".{os.getpid()}.tmp")
    tmp.write_text(text, encoding="utf-8")
    os.replace(tmp, path)


def save_thread(th: CivilThread) -> None:
    _DIR.mkdir(parents=True, exist_ok=True)
    th.updated_at = time.time()
    _atomic_write(_path(th.thread_id), json.dumps(th.to_dict(), ensure_ascii=False, indent=2))


def _messages_path(thread_id: str) -> Path:
    p = _path(thread_id)
    return p.with_name(p.stem + ".messages.jsonl")


def append_message(thread_id: str, role: str, content: str, **extra: Any) -> None:
    """Append one transcript line; the thread JSON keeps only a counter so listings stay small."""
    if not thread_id:
        return
    _DIR.mkdir(parents=True, exist_ok=True)
    row = {"role": role, "content": content, "ts": time.time(), **extra}
    with _LOCK:
        with _messages_path(thread_id).open("a", encoding="utf-8") as fh:
            fh.write(json.dumps(row, ensure_ascii=False) + "\n")
    th = load_thread(thread_id)
    if th:
        th.n_messages += 1
        if role == "user" and (not th.title or th.title == "新对话"):
            th.title = content.replace("\n", " ")[:40] or th.title
        save_thread(th)


def load_messages(thread_id: str, limit: int = MAX_MESSAGES) -> List[Dict[str, Any]]:
    p = _messages_path(thread_id)
    if not p.is_file():
        return []
    rows: List[Dict[str, Any]] = []
    for line in p.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            row = json.loads(line)
        except json.JSONDecodeError:
            continue
        if isinstance(row, dict) and row.get("role") in {"user", "assistant"}:
            rows.append(row)
    return rows[-limit:]


def delete_thread(thread_id: str) -> bool:
    ok = False
    for p in (_path(thread_id), _messages_path(thread_id)):
        try:
            p.unlink()
            ok = True
        except FileNotFoundError:
            continue
    _FUTS.pop(thread_id, None)
    _CANCEL.discard(thread_id)
    return ok


def request_cancel(thread_id: str) -> Dict[str, Any]:
    """Cooperative cancel: background futures not yet started are dropped; running work checks
    cancel_requested() between steps (the /api/chat stream does this after every LLM step)."""
    th = load_thread(thread_id)
    if not th:
        return {"ok": False, "error": "unknown thread"}
    _CANCEL.add(thread_id)
    fut = _FUTS.get(thread_id)
    dropped = bool(fut and fut.cancel())
    th.cancel_requested = True
    if dropped or th.state in {"idle", "stale"}:
        th.state = "cancelled"
    save_thread(th)
    return {"ok": True, "thread_id": thread_id, "state": th.state, "dropped": dropped}


def cancel_requested(thread_id: str) -> bool:
    return bool(thread_id) and thread_id in _CANCEL


def clear_cancel(thread_id: str) -> None:
    _CANCEL.discard(thread_id)


def load_thread(thread_id: str) -> Optional[CivilThread]:
    p = _path(thread_id)
    if not p.is_file():
        return None
    try:
        raw = json.loads(p.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None
    if not isinstance(raw, dict):
        return None
    return CivilThread(
        thread_id=str(raw.get("thread_id") or thread_id),
        session_id=str(raw.get("session_id") or thread_id),
        title=str(raw.get("title") or ""),
        skill=str(raw.get("skill") or ""),
        state=str(raw.get("state") or "idle"),
        confirm=bool(raw.get("confirm")),
        last_text=str(raw.get("last_text") or ""),
        last_reply=str(raw.get("last_reply") or ""),
        hitl_pending=bool(raw.get("hitl_pending")),
        wrote=bool(raw.get("wrote")),
        artifacts=list(raw.get("artifacts") or []),
        error=str(raw.get("error") or ""),
        created_at=float(raw.get("created_at") or 0),
        updated_at=float(raw.get("updated_at") or 0),
        cancel_requested=bool(raw.get("cancel_requested")),
        n_messages=int(raw.get("n_messages") or 0),
    )


def list_threads() -> List[CivilThread]:
    if not _DIR.is_dir():
        return []
    out: List[CivilThread] = []
    for p in sorted(_DIR.glob("*.json"), key=lambda x: x.stat().st_mtime, reverse=True):
        if p.name.endswith(".tmp"):
            continue
        th = load_thread(p.stem)
        if not th:
            continue
        fut = _FUTS.get(th.thread_id)
        if th.state == "running" and not (fut and not fut.done()) and th.updated_at < _BOOT:
            # left over from a previous process: nothing can finish it any more
            th.state = "stale"
            th.error = th.error or "工作台重启，任务没有跑完；请重新发送"
            save_thread(th)
        out.append(th)
    return out


def new_thread(title: str = "", *, confirm: bool = False) -> CivilThread:
    tid = f"t-{uuid4().hex[:8]}"
    th = CivilThread(
        thread_id=tid,
        session_id=tid,
        title=(title or "新对话").strip()[:80],
        confirm=confirm,
    )
    save_thread(th)
    return th


def _run_on_thread(thread_id: str, text: str, *, skill: str, confirm: bool) -> Dict[str, Any]:
    from packing_assistant.runtime.agent_loop import run_agent

    th = load_thread(thread_id) or new_thread()
    th.state = "running"
    th.last_text = text
    if not th.title or th.title == "新对话":
        th.title = (text or "").replace("\n", " ")[:40] or th.title
    save_thread(th)
    try:
        out = run_agent(
            text,
            session_id=th.session_id,
            expert_id=skill,
            p0_confirmed=confirm or th.confirm,
        )
    except Exception as exc:  # noqa: BLE001
        th.state = "failed"
        th.error = str(exc)
        th.last_reply = str(exc)
        save_thread(th)
        return {"ok": False, "error": str(exc), "thread_id": thread_id}
    th.skill = str(out.get("skill") or out.get("expert_id") or th.skill)
    th.last_reply = str(out.get("reply") or "")
    th.hitl_pending = bool(out.get("hitl_pending"))
    th.wrote = bool(out.get("wrote"))
    arts = out.get("artifacts") or out.get("files") or []
    th.artifacts = [str(a) for a in arts]
    th.state = "waiting_hitl" if th.hitl_pending else ("done" if out.get("ok") else "failed")
    th.error = str(out.get("error") or out.get("error_code") or "")
    save_thread(th)
    out = dict(out)
    out["thread_id"] = thread_id
    return out


def run_on_thread(
    thread_id: str,
    text: str,
    *,
    skill: str = "",
    confirm: bool = False,
    background: bool = False,
) -> Dict[str, Any]:
    if background:
        with _LOCK:
            fut = _pool().submit(_run_on_thread, thread_id, text, skill=skill, confirm=confirm)
            _FUTS[thread_id] = fut
        return {
            "ok": True,
            "background": True,
            "thread_id": thread_id,
            "state": "running",
        }
    return _run_on_thread(thread_id, text, skill=skill, confirm=confirm)


def spawn(text: str, *, skill: str = "", confirm: bool = False, title: str = "") -> Dict[str, Any]:
    th = new_thread(title or text, confirm=confirm)
    return run_on_thread(th.thread_id, text, skill=skill, confirm=confirm, background=True)


def thread_status(thread_id: str) -> Dict[str, Any]:
    th = load_thread(thread_id)
    if not th:
        return {"ok": False, "error": "unknown thread"}
    fut = _FUTS.get(thread_id)
    running = bool(fut and not fut.done())
    if running:
        th.state = "running"
    elif th.state == "running" and th.updated_at < _BOOT:
        th.state = "stale"
    return {"ok": True, **th.to_dict(), "running": running}

File: workbench/src/threads.rs
//! Conversation threads, shared on disk with the Python workbench.
//!
//! Layout (identical to `packing_assistant/runtime/threads.py`):
//!   <out_root>/_threads/<thread_id>.json            thread record
//!   <out_root>/_threads/<thread_id>.messages.jsonl  transcript, one row per turn
//! Either backend can open a thread the other one wrote.

use crate::config::Paths;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashSet;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::{Mutex, OnceLock};
use std::time::{SystemTime, UNIX_EPOCH};

pub const MAX_MESSAGES: usize = 400;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CivilThread {
    pub thread_id: String,
    pub session_id: String,
    #[serde(default)]
    pub title: String,
    #[serde(default)]
    pub skill: String,
    #[serde(default = "idle")]
    pub state: String,
    #[serde(default)]
    pub confirm: bool,
    #[serde(default)]
    pub last_text: String,
    #[serde(default)]
    pub last_reply: String,
    #[serde(default)]
    pub hitl_pending: bool,
    #[serde(default)]
    pub wrote: bool,
    #[serde(default)]
    pub artifacts: Vec<String>,
    #[serde(default)]
    pub error: String,
    #[serde(default)]
    pub created_at: f64,
    #[serde(default)]
    pub updated_at: f64,
    #[serde(default)]
    pub cancel_requested: bool,
    #[serde(default)]
    pub n_messages: u64,
}

fn idle() -> String {
    "idle".into()
}

pub fn now() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0)
}

fn boot_time() -> f64 {
    static BOOT: OnceLock<f64> = OnceLock::new();
    *BOOT.get_or_init(now)
}

fn cancel_set() -> &'static Mutex<HashSet<String>> {
    static SET: OnceLock<Mutex<HashSet<String>>> = OnceLock::new();
    SET.get_or_init(|| Mutex::new(HashSet::new()))
}

fn running_set() -> &'static Mutex<HashSet<String>> {
    static SET: OnceLock<Mutex<HashSet<String>>> = OnceLock::new();
    SET.get_or_init(|| Mutex::new(HashSet::new()))
}

pub fn dir(paths: &Paths) -> PathBuf {
    paths.out_root.join("_threads")
}

fn safe_id(thread_id: &str) -> String {
    let s: String = thread_id
        .chars()
        .map(|c| if c.is_ascii_alphanumeric() || c == '-' || c == '_' { c } else { '_' })
        .take(40)
        .collect();
    if s.is_empty() {
        "t".into()
    } else {
        s
    }
}

pub fn path(paths: &Paths, thread_id: &str) -> PathBuf {
    dir(paths).join(format!("{}.json", safe_id(thread_id)))
}

fn messages_path(paths: &Paths, thread_id: &str) -> PathBuf {
    dir(paths).join(format!("{}.messages.jsonl", safe_id(thread_id)))
}

fn atomic_write(target: &Path, text: &str) -> Result<(), String> {
    if let Some(parent) = target.parent() {
        fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    let tmp = target.with_extension(format!("json.{}.tmp", std::process::id()));
    fs::write(&tmp, text).map_err(|e| e.to_string())?;
    fs::rename(&tmp, target).map_err(|e| e.to_string())
}

pub fn save(paths: &Paths, th: &mut CivilThread) -> Result<(), String> {
    th.updated_at = now();
    let text = serde_json::to_string_pretty(th).map_err(|e| e.to_string())?;
    atomic_write(&path(paths, &th.thread_id), &text)
}

pub fn load(paths: &Paths, thread_id: &str) -> Option<CivilThread> {
    let raw = fs::read_to_string(path(paths, thread_id)).ok()?;
    let mut th: CivilThread = serde_json::from_str(&raw).ok()?;
    if th.thread_id.is_empty() {
        th.thread_id = thread_id.to_string();
    }
    if th.session_id.is_empty() {
        th.session_id = th.thread_id.clone();
    }
    Some(th)
}

pub fn new_thread(paths: &Paths, title: &str, confirm: bool) -> Result<CivilThread, String> {
    let tid = format!("t-{}", &uuid::Uuid::new_v4().simple().to_string()[..8]);
    let title = title.trim();
    let mut th = CivilThread {
        thread_id: tid.clone(),
        session_id: tid,
        title: if title.is_empty() { "新对话".into() } else { title.chars().take(80).collect() },
        skill: String::new(),
        state: "idle".into(),
        confirm,
        last_text: String::new(),
        last_reply: String::new(),
        hitl_pending: false,
        wrote: false,
        artifacts: vec![],
        error: String::new(),
        created_at: now(),
        updated_at: now(),
        cancel_requested: false,
        n_messages: 0,
    };
    save(paths, &mut th)?;
    Ok(th)
}

pub fn mark_running(thread_id: &str, on: bool) {
    let mut set = running_set().lock().unwrap_or_else(|e| e.into_inner());
    if on {
        set.insert(thread_id.to_string());
    } else {
        set.remove(thread_id);
    }
}

pub fn is_running(thread_id: &str) -> bool {
    running_set()
        .lock()
        .unwrap_or_else(|e| e.into_inner())
        .contains(thread_id)
}

/// Threads newest first. A `running` record older than this process with no live task
/// behind it is reported (and persisted) as `stale`: nothing can finish it any more.
pub fn list(paths: &Paths) -> Vec<CivilThread> {
    let Ok(rd) = fs::read_dir(dir(paths)) else {
        return vec![];
    };
    let mut out: Vec<CivilThread> = Vec::new();
    for ent in rd.flatten() {
        let p = ent.path();
        let name = p.file_name().and_then(|s| s.to_str()).unwrap_or("");
        if !name.ends_with(".json") || name.contains(".tmp") {
            continue;
        }
        let stem = name.trim_end_matches(".json");
        let Some(mut th) = load(paths, stem) else {
            continue;
        };
        if th.state == "running" && !is_running(&th.thread_id) && th.updated_at < boot_time() {
            th.state = "stale".into();
            if th.error.is_empty() {
                th.error = "工作台重启，任务没有跑完；请重新发送".into();
            }
            let _ = save(paths, &mut th);
        }
        out.push(th);
    }
    out.sort_by(|a, b| b.updated_at.partial_cmp(&a.updated_at).unwrap_or(std::cmp::Ordering::Equal));
    out
}

pub fn append_message(paths: &Paths, thread_id: &str, role: &str, content: &str, extra: Value) -> Result<(), String> {
    if thread_id.is_empty() {
        return Ok(());
    }
    let mut row = json!({"role": role, "content": content, "ts": now()});
    if let Some(map) = extra.as_object() {
        for (k, v) in map {
            row[k] = v.clone();
        }
    }
    let p = messages_path(paths, thread_id);
    if let Some(parent) = p.parent() {
        fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    use std::io::Write;
    let mut fh = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(&p)
        .map_err(|e| e.to_string())?;
    writeln!(fh, "{}", row).map_err(|e| e.to_string())?;
    if let Some(mut th) = load(paths, thread_id) {
        th.n_messages += 1;
        if role == "user" && (th.title.is_empty() || th.title == "新对话") {
            let t: String = content.replace('\n', " ").chars().take(40).collect();
            if !t.is_empty() {
                th.title = t;
            }
        }
        save(paths, &mut th)?;
    }
    Ok(())
}

pub fn load_messages(paths: &Paths, thread_id: &str, limit: usize) -> Vec<Value> {
    let Ok(raw) = fs::read_to_string(messages_path(paths, thread_id)) else {
        return vec![];
    };
    let mut rows: Vec<Value> = raw
        .lines()
        .filter_map(|l| serde_json::from_str::<Value>(l.trim()).ok())
        .filter(|v| matches!(v.get("role").and_then(|r| r.as_str()), Some("user") | Some("assistant")))
        .collect();
    let keep = limit.clamp(1, 2000);
    if rows.len() > keep {
        rows = rows.split_off(rows.len() - keep);
    }
    rows
}

pub fn delete(paths: &Paths, thread_id: &str) -> bool {
    let mut ok = false;
    for p in [path(paths, thread_id), messages_path(paths, thread_id)] {
        if fs::remove_file(p).is_ok() {
            ok = true;
        }
    }
    clear_cancel(thread_id);
    mark_running(thread_id, false);
    ok
}

pub fn request_cancel(paths: &Paths, thread_id: &str) -> Option<Value> {
    let mut th = load(paths, thread_id)?;
    cancel_set()
        .lock()
        .unwrap_or_else(|e| e.into_inner())
        .insert(thread_id.to_string());
    th.cancel_requested = true;
    if !is_running(thread_id) && matches!(th.state.as_str(), "idle" | "stale") {
        th.state = "cancelled".into();
    }
    let _ = save(paths, &mut th);
    Some(json!({"ok": true, "thread_id": thread_id, "state": th.state, "dropped": false}))
}

pub fn cancel_requested(thread_id: &str) -> bool {
    !thread_id.is_empty()
        && cancel_set()
            .lock()
            .unwrap_or_else(|e| e.into_inner())
            .contains(thread_id)
}

pub fn clear_cancel(thread_id: &str) {
    cancel_set()
        .lock()
        .unwrap_or_else(|e| e.into_inner())
        .remove(thread_id);
}

/// Every deliverable under <out_root>/<session>/<expert>/ — survives a broken stream.
pub fn session_files(paths: &Paths, session_id: &str) -> Vec<Value> {
    let sid: String = session_id
        .chars()
        .filter(|c| c.is_ascii_alphanumeric() || *c == '-' || *c == '_')
        .take(32)
        .collect();
    if sid.len() < 4 {
        return vec![];
    }
    let root = paths.out_root.join(&sid);
    let Ok(rd) = fs::read_dir(&root) else {
        return vec![];
    };
    let mut rows: Vec<(f64, Value)> = Vec::new();
    for expert_dir in rd.flatten() {
        let ep = expert_dir.path();
        if !ep.is_dir() {
            continue;
        }
        let expert = ep.file_name().and_then(|s| s.to_str()).unwrap_or("").to_string();
        let Ok(files) = fs::read_dir(&ep) else {
            continue;
        };
        for f in files.flatten() {
            let fp = f.path();
            let Ok(meta) = fp.metadata() else {
                continue;
            };
            if !meta.is_file() {
                continue;
            }
            let mtime = meta
                .modified()
                .ok()
                .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
                .map(|d| d.as_secs_f64())
                .unwrap_or(0.0);
            rows.push((
                mtime,
                json!({
                    "expert": expert,
                    "name": fp.file_name().and_then(|s| s.to_str()).unwrap_or(""),
                    "path": fp.to_string_lossy(),
                    "bytes": meta.len(),
                    "mtime": mtime,
                }),
            ));
        }
    }
    rows.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap_or(std::cmp::Ordering::Equal));
    rows.into_iter().map(|(_, v)| v).collect()
}

pub fn status(paths: &Paths, thread_id: &str) -> Option<Value> {
    let mut th = load(paths, thread_id)?;
    let running = is_running(thread_id);
    if running {
        th.state = "running".into();
    } else if th.state == "running" && th.updated_at < boot_time() {
        th.state = "stale".into();
    }
    let mut v = serde_json::to_value(&th).ok()?;
    v["ok"] = json!(true);
    v["running"] = json!(running);
    Some(v)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn paths() -> Paths {
        let tmp = std::env::temp_dir().join(format!("cb-threads-{}", uuid::Uuid::new_v4().simple()));
        fs::create_dir_all(tmp.join("out")).unwrap();
        Paths::from_demo(tmp)
    }

    #[test]
    fn transcript_roundtrip_and_title() {
        let p = paths();
        let th = new_thread(&p, "", false).unwrap();
        assert_eq!(th.title, "新对话");
        append_message(&p, &th.thread_id, "user", "写一份临边提纲\n第二行", json!({"experts": ["construction"]})).unwrap();
        append_message(&p, &th.thread_id, "assistant", "好的", json!({"expert": "construction"})).unwrap();
        let rows = load_messages(&p, &th.thread_id, 400);
        assert_eq!(rows.len(), 2);
        assert_eq!(rows[1]["expert"], "construction");
        let th2 = load(&p, &th.thread_id).unwrap();
        assert_eq!(th2.n_messages, 2);
        assert_eq!(th2.title, "写一份临边提纲 第二行");
        assert!(delete(&p, &th.thread_id));
        assert!(load(&p, &th.thread_id).is_none());
    }

    #[test]
    fn stale_and_cancel() {
        let p = paths();
        let mut th = new_thread(&p, "orphan", false).unwrap();
        th.state = "running".into();
        save(&p, &mut th).unwrap();
        // pretend it was written before this process started
        th.updated_at = boot_time() - 10.0;
        atomic_write(&path(&p, &th.thread_id), &serde_json::to_string(&th).unwrap()).unwrap();
        let rows = list(&p);
        assert_eq!(rows[0].state, "stale");

        let th2 = new_thread(&p, "cancel-me", false).unwrap();
        let got = request_cancel(&p, &th2.thread_id).unwrap();
        assert_eq!(got["state"], "cancelled");
        assert!(cancel_requested(&th2.thread_id));
        clear_cancel(&th2.thread_id);
        assert!(!cancel_requested(&th2.thread_id));
        assert!(request_cancel(&p, "nope").is_none());
    }

    #[test]
    fn same_layout_as_python() {
        let p = paths();
        let th = new_thread(&p, "x", true).unwrap();
        let raw = fs::read_to_string(path(&p, &th.thread_id)).unwrap();
        let v: Value = serde_json::from_str(&raw).unwrap();
        for key in [
            "thread_id", "session_id", "title", "skill", "state", "confirm", "last_text", "last_reply",
            "hitl_pending", "wrote", "artifacts", "error", "created_at", "updated_at", "cancel_requested", "n_messages",
        ] {
            assert!(v.get(key).is_some(), "missing {key}");
        }
    }
}
